gunrock/app/bc/bc_enactor.cuh

```diff
@@ -246,124 +232,116 @@ struct BCForwardIterationLoop : public IterationLoopBase
+        auto    &deltas             =   data_slice.deltas;
+        auto    &labels             =   data_slice.labels;
+        auto     src_node           =   data_slice.src_node;
+        auto     num_vertices       =   graph.nodes;
 
         auto advance_op = [
-            labels, deltas, bc_values, iteration, src_node, sigmas
+            labels, deltas, bc_values, iteration, src_node, sigmas, num_vertices
         ] __host__ __device__ (
             const VertexT &src, VertexT &dest, const SizeT &edge_id,
             const VertexT &input_item, const SizeT &input_pos,
             SizeT &output_pos) -> bool
         {
-            // printf("backward advance_op");
-            // printf("iteration=%d \n", iteration);
-
             VertexT s_label = Load<cub::LOAD_CG>(labels + src);
             VertexT d_label = Load<cub::LOAD_CG>(labels + dest);
 
-            if(iteration == 0) {
+            if (iteration == 0) {
                 return (d_label == s_label + 1);
             } else {
                 if (d_label == s_label + 1) {
-                    if (src == src_node[0]) return true; // !! Is this right?
-                    
+                    if (src == src_node) return true; // !! Is this right? YC: it's right
+
                     ValueT from_sigma = Load<cub::LOAD_CG>(sigmas + src);
                     ValueT to_sigma   = Load<cub::LOAD_CG>(sigmas + dest);
                     ValueT to_delta   = Load<cub::LOAD_CG>(deltas + dest);
                     ValueT result     = from_sigma / to_sigma * (1.0 + to_delta);
-                    
+
                     //Accumulate bc value
-                    {
-                        ValueT old_delta    = atomicAdd(deltas + src, result);
-                        ValueT old_bc_value = atomicAdd(bc_values + src, result);
-                    }
+                    ValueT old_delta    = atomicAdd(deltas + src, result);
+                    ValueT old_bc_value = atomicAdd(bc_values + src, result);
                     return true;
                 } else {
                     return false;
-                }               
+                }
             }
         };
 
         auto filter_op = [
-            labels
+            labels, iteration
         ] __host__ __device__ (
             const VertexT &src, VertexT &dest, const SizeT &edge_id,
             const VertexT &input_item, const SizeT &input_pos,
             SizeT &output_pos) -> bool
         {
-            // printf("backward filter_op");
-            return labels + dest == 0;
+            return labels[dest] == 0;
         };
-        
-        // Call the advance operator, using the advance operation
-        // TODO: modify the operator callers according to algorithmic needs,
-        //       this example only uses an advance + a filter, with
-        //       possible optimization to fuze the two kernels.
-        //       Define more operations (i.e. lambdas) if needed
+
+        frontier.queue_reset = true;
```

Comment:
What does this do?  Are these kinds of `frontier` operations documented somewhere?

Reply:
This particular line is the solution to a bug in the backward iteration that the first thread of a block in the advance operator didn't get the correct frontier size.

`queue_reset` tells the operators to use frontier size provided via `queue_length`, rather than the one resulted from the pervious operator. It's from the old API, and I should make it more clear somewhere.